fix terrible bug and inline GC for expiring cache

mikedanese opened the pull request:
I fudged cleanup of this comment:

https://github.com/kubernetes/kubernetes/pull/84424#discussion_r345346041

It needs to be the opposite of this:

https://github.com/kubernetes/kubernetes/blob/f2469a7a2c75a78b8c0b0514652511d105d2b2e7/staging/src/k8s.io/apimachinery/pkg/util/cache/expiring.go#L75

I've made the expiration test check cache size rather than just calling Get.

This change also inlines GC, removing the background goroutine. This appears to make virtually no difference in performance. I'm happy to separate out the bug fix as well.

Before (including the fix) and after (including fix plus GC inlining) benchmarks for 1s and 10s:

#### -benchtime=1s
Mostly loading.
```
$ benchstat before.1s.log after.1s.log 
name                                                  old time/op     new time/op     delta
CachedTokenAuthenticator/tokens=100_threads=1-60          661ns ± 6%      655ns ±12%     ~     (p=0.797 n=9+10)
CachedTokenAuthenticator/tokens=100_threads=16-60         594ns ±15%      502ns ±17%  -15.43%  (p=0.002 n=10+9)
CachedTokenAuthenticator/tokens=100_threads=256-60        649ns ±35%      686ns ±20%     ~     (p=0.387 n=9+9)
CachedTokenAuthenticator/tokens=500_threads=1-60          376ns ± 2%      387ns ± 2%   +2.98%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=500_threads=16-60         430ns ±17%      408ns ±14%     ~     (p=0.160 n=10+10)
CachedTokenAuthenticator/tokens=500_threads=256-60        843ns ±21%      674ns ±18%  -20.08%  (p=0.002 n=9+9)
CachedTokenAuthenticator/tokens=2500_threads=1-60         364ns ± 1%      361ns ± 1%     ~     (p=0.105 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=16-60        447ns ±13%      432ns ±13%     ~     (p=0.342 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=256-60      1.22µs ±25%     1.23µs ±30%     ~     (p=0.853 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=1-60        446ns ± 1%      407ns ± 1%   -8.82%  (p=0.000 n=8+9)
CachedTokenAuthenticator/tokens=12500_threads=16-60       602ns ± 8%      600ns ± 7%     ~     (p=0.838 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=256-60     12.1µs ± 8%     12.7µs ± 2%     ~     (p=0.278 n=10+9)
CachedTokenAuthenticator/tokens=62500_threads=1-60       18.0µs ± 0%     17.9µs ± 0%   -0.17%  (p=0.021 n=9+10)
CachedTokenAuthenticator/tokens=62500_threads=16-60      18.8µs ± 0%     18.8µs ± 0%     ~     (p=0.912 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=256-60     25.0µs ± 1%     25.1µs ± 1%     ~     (p=0.684 n=10+10)

name                                                  old lookups/op  new lookups/op  delta
CachedTokenAuthenticator/tokens=100_threads=1-60           0.00 ±12%       0.00 ±14%     ~     (p=0.754 n=10+10)
CachedTokenAuthenticator/tokens=100_threads=16-60          0.00 ±49%      0.00 ±103%     ~     (p=0.725 n=10+10)
CachedTokenAuthenticator/tokens=100_threads=256-60         0.01 ±46%       0.01 ±50%     ~     (p=0.450 n=9+9)
CachedTokenAuthenticator/tokens=500_threads=1-60           0.00 ± 3%       0.00 ± 5%     ~     (p=0.090 n=9+10)
CachedTokenAuthenticator/tokens=500_threads=16-60          0.00 ±59%       0.00 ±12%  -34.78%  (p=0.004 n=10+10)
CachedTokenAuthenticator/tokens=500_threads=256-60         0.02 ±20%       0.02 ±68%  -30.19%  (p=0.007 n=8+10)
CachedTokenAuthenticator/tokens=2500_threads=1-60          0.00 ± 2%       0.00 ± 3%   -5.10%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=16-60         0.00 ±24%       0.00 ±16%     ~     (p=0.109 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=256-60        0.03 ±25%       0.04 ±27%     ~     (p=0.853 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=1-60         0.01 ± 1%       0.01 ± 1%  -15.13%  (p=0.000 n=8+9)
CachedTokenAuthenticator/tokens=12500_threads=16-60        0.01 ± 7%       0.01 ±12%     ~     (p=0.221 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=256-60       0.40 ± 7%       0.42 ± 1%     ~     (p=0.055 n=10+9)
CachedTokenAuthenticator/tokens=62500_threads=1-60         0.65 ± 0%       0.65 ± 0%   -0.12%  (p=0.029 n=7+10)
CachedTokenAuthenticator/tokens=62500_threads=16-60        0.68 ± 0%       0.68 ± 0%     ~     (p=0.735 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=256-60       0.88 ± 1%       0.88 ± 1%     ~     (p=0.929 n=10+10)

name                                                  old alloc/op    new alloc/op    delta
CachedTokenAuthenticator/tokens=100_threads=1-60           300B ± 0%       300B ± 0%     ~     (p=1.000 n=10+10)
CachedTokenAuthenticator/tokens=100_threads=16-60          302B ± 0%       302B ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=100_threads=256-60         345B ± 5%       349B ± 5%     ~     (p=0.266 n=9+9)
CachedTokenAuthenticator/tokens=500_threads=1-60           297B ± 0%       297B ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=16-60          299B ± 0%       299B ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=256-60         354B ± 3%       344B ± 5%   -2.86%  (p=0.006 n=9+9)
CachedTokenAuthenticator/tokens=2500_threads=1-60          297B ± 0%       297B ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=16-60         298B ± 0%       299B ± 0%   +0.17%  (p=0.033 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=256-60        368B ±10%       397B ± 6%   +7.74%  (p=0.006 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=1-60         298B ± 0%       266B ± 0%  -10.74%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=16-60        302B ± 0%       302B ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=12500_threads=256-60     1.39kB ± 8%     1.44kB ± 2%     ~     (p=0.189 n=10+9)
CachedTokenAuthenticator/tokens=62500_threads=1-60         566B ± 0%       566B ± 0%     ~     (p=0.549 n=8+10)
CachedTokenAuthenticator/tokens=62500_threads=16-60        663B ± 0%       663B ± 0%     ~     (p=0.452 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=256-60     2.66kB ± 9%     2.69kB ± 4%     ~     (p=0.811 n=10+10)

name                                                  old allocs/op   new allocs/op   delta
CachedTokenAuthenticator/tokens=100_threads=1-60           11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=100_threads=16-60          11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=100_threads=256-60         11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=1-60           11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=16-60          11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=256-60         11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=1-60          11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=16-60         11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=256-60        10.4 ± 6%       11.4 ± 5%   +9.62%  (p=0.002 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=1-60         11.0 ± 0%       10.0 ± 0%   -9.09%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=16-60        11.0 ± 0%       11.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=12500_threads=256-60       14.0 ± 0%       14.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=62500_threads=1-60         15.0 ± 0%       15.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=62500_threads=16-60        15.0 ± 0%       15.0 ± 0%     ~     (all equal)
CachedTokenAuthenticator/tokens=62500_threads=256-60       16.0 ± 0%       17.0 ± 0%   +6.25%  (p=0.000 n=10+10)
```

#### -benchtime=10s
Long enough to test caching and GC.
```
$ benchstat before.10s.log after.10s.log 
name                                                  old time/op     new time/op     delta
CachedTokenAuthenticator/tokens=100_threads=1-60          650ns ± 5%      632ns ±10%    ~     (p=0.690 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=16-60         513ns ± 8%      500ns ± 6%    ~     (p=0.516 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=256-60        744ns ±15%      701ns ±14%    ~     (p=0.548 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=1-60          374ns ± 1%      373ns ± 0%    ~     (p=0.968 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=16-60         428ns ± 5%      421ns ± 4%    ~     (p=0.421 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=256-60        814ns ±11%      855ns ±11%    ~     (p=0.460 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=1-60         346ns ± 1%      352ns ± 1%  +1.79%  (p=0.016 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=16-60        418ns ± 7%      415ns ±16%    ~     (p=0.278 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=256-60       980ns ±10%      946ns ± 7%    ~     (p=0.548 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=1-60        309ns ± 1%      309ns ± 1%    ~     (p=0.913 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=16-60       494ns ± 9%      509ns ± 7%    ~     (p=0.516 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=256-60     1.60µs ± 2%     1.57µs ± 1%  -1.89%  (p=0.040 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=1-60        609ns ± 0%      616ns ± 0%  +1.12%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=16-60       710ns ± 0%      742ns ± 1%  +4.54%  (p=0.029 n=4+4)
CachedTokenAuthenticator/tokens=62500_threads=256-60     4.52µs ± 2%     4.51µs ± 1%    ~     (p=1.000 n=5+5)

name                                                  old lookups/op  new lookups/op  delta
CachedTokenAuthenticator/tokens=100_threads=1-60           0.00 ±10%       0.00 ± 7%    ~     (p=0.310 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=16-60          0.00 ±29%       0.00 ±28%    ~     (p=0.421 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=256-60         0.02 ±24%       0.02 ±24%    ~     (p=0.548 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=1-60           0.00 ± 2%       0.00 ± 6%    ~     (p=0.690 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=16-60          0.00 ± 9%       0.00 ± 7%    ~     (p=0.151 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=256-60         0.02 ±17%       0.03 ±14%    ~     (p=0.310 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=1-60          0.00 ± 1%       0.00 ± 1%  +1.78%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=16-60         0.00 ±21%       0.00 ±23%    ~     (p=0.690 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=256-60        0.03 ±12%       0.03 ± 7%    ~     (p=0.595 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=1-60         0.00 ± 1%       0.00 ± 1%    ~     (p=0.841 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=16-60        0.01 ±16%       0.01 ±12%    ~     (p=0.500 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=256-60       0.05 ± 2%       0.05 ± 1%    ~     (p=0.151 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=1-60         0.02 ± 0%       0.02 ± 0%  +0.92%  (p=0.016 n=4+5)
CachedTokenAuthenticator/tokens=62500_threads=16-60        0.02 ± 1%       0.02 ±11%    ~     (p=0.746 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=256-60       0.16 ± 2%       0.16 ± 1%    ~     (p=0.873 n=5+5)

name                                                  old alloc/op    new alloc/op    delta
CachedTokenAuthenticator/tokens=100_threads=1-60           299B ± 0%       299B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=100_threads=16-60          293B ± 0%       293B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=100_threads=256-60         303B ± 0%       304B ± 0%    ~     (p=0.556 n=5+4)
CachedTokenAuthenticator/tokens=500_threads=1-60           296B ± 0%       296B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=16-60          296B ± 0%       296B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=256-60         305B ± 0%       305B ± 0%    ~     (p=0.444 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=1-60          297B ± 0%       297B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=16-60         297B ± 0%       297B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=256-60        307B ± 0%       306B ± 0%    ~     (p=0.095 n=4+5)
CachedTokenAuthenticator/tokens=12500_threads=1-60         264B ± 0%       264B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=12500_threads=16-60        297B ± 0%       297B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=12500_threads=256-60       315B ± 0%       314B ± 0%    ~     (p=0.095 n=5+4)
CachedTokenAuthenticator/tokens=62500_threads=1-60         299B ± 0%       299B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=62500_threads=16-60        300B ± 0%       300B ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=62500_threads=256-60       349B ± 0%       349B ± 0%    ~     (p=0.444 n=5+5)

name                                                  old allocs/op   new allocs/op   delta
CachedTokenAuthenticator/tokens=100_threads=1-60           11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=100_threads=16-60          11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=100_threads=256-60         11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=1-60           11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=16-60          11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=500_threads=256-60         11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=1-60          11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=16-60         11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=2500_threads=256-60        11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=12500_threads=1-60         10.0 ± 0%       10.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=12500_threads=16-60        11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=12500_threads=256-60       12.0 ± 0%       12.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=62500_threads=1-60         11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=62500_threads=16-60        11.0 ± 0%       11.0 ± 0%    ~     (all equal)
CachedTokenAuthenticator/tokens=62500_threads=256-60       12.0 ± 0%       12.0 ± 0%    ~     (all equal)
```

/kind bug
/sig auth
Fixes: #85317

```release-note
NONE
```

lavalamp commented:
huh, I'm surprised the goroutine removal did so little.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/cache/expiring.go:
```diff
@@ -164,7 +147,7 @@ func (c *Expiring) gc() {
 		// heap.Pop() swaps the first entry with the last entry of the heap, then
 		// calls (*expiringHeap).Pop() which returns the last element.
-		if len(c.heap) == 0 || now.After(c.heap[0].expiry) {
+		if len(c.heap) == 0 || !now.After(c.heap[0].expiry) {
```
How did this pass the tests previously?

mikedanese replied:
We were checking Get() which drops expired entries but not checking cache size.

enj replied:
`c.heap[0].expiry.After(now)` lines up with the comment better IMO.

lavalamp commented:
Can you add a test that verifies that a call to gc removes things it should and doesn't remove things it shouldn't?

lavalamp commented:
Also how do you generate the awesome stats from the benchmarks?

mikedanese commented:
> huh, I'm surprised the goroutine removal did so little.

I've noticed that GC is actually very cheap vs the time spent in Set() (even before the bug was introduced). I'm guessing that the minimal time spent in GC spreads pretty evenly when inlined. Will validate.

> Can you add a test that verifies that a call to gc removes things it should and doesn't remove things it shouldn't?

Yup, will do.

> Also how do you generate the awesome stats from the benchmarks?

Benchmarks run on a c2-standard-60:

```
[master]$ go test -c -o old ./vendor/k8s.io/apiserver/pkg/authentication/token/cache
[expcache]$ go test -c -o new ./vendor/k8s.io/apiserver/pkg/authentication/token/cache
```

```
#!/bin/bash

set -o errexit
set -o xtrace
set -o nounset
set -o pipefail

SHARED="-test.v -test.benchmem -test.bench='CachedTokenAuthenticator/.*' -test.run='^$' -test.timeout=1h"
SHORT="${SHARED} -test.count=10 -test.benchtime=1s"
LONG="${SHARED} -test.count=5 -test.benchtime=10s"

gcloud compute scp --zone us-central1-c old new perf:~/

gcloud compute ssh  perf --zone us-central1-c \
  --command="./old ${SHORT}" \
  | tee before.1s.log

gcloud compute ssh  perf --zone us-central1-c \
  --command="./new ${SHORT}" \
  | tee after.1s.log

bencstat before.1s.log after.1s.log | tee 1s.diff.log

gcloud compute ssh  perf --zone us-central1-c \
  --command="./old ${LONG}" \
  | tee before.10s.log

gcloud compute ssh  perf --zone us-central1-c \
  --command="./new ${LONG}" \
  | tee after.10s.log

bencstat before.10s.log after.10s.log | tee 1s.diff.log
```

https://godoc.org/golang.org/x/perf/cmd/benchstat

enj reviewed (commented):
Sameish comment that led to this bug...

lavalamp commented:
> I've noticed that GC is actually very cheap vs the time spent in Set() (even before the bug was introduced). I'm guessing that the minimal time spent in GC spreads pretty evenly when inlined. Will validate.

Yeah, I was hoping for an improvement, but even though it's a no-op removing the goroutine is still a large code complexity improvement, I think.

Do you want to add a test to this PR or in a followup?

Thanks for the stats write-up, I'll be referring to that in the future.

mikedanese commented:
> Do you want to add a test to this PR or in a followup?

No, I can do it now.

mikedanese commented:
Also here's LRU cache compared to fixed Expiring with no inline GC. Had to make sure I wasn't going crazy.

#### -benchtime=1s
Mostly loading.
```
name                                                  old time/op     new time/op     delta
CachedTokenAuthenticator/tokens=100_threads=1-60         1.52µs ±12%     0.66µs ± 6%  -56.37%  (p=0.000 n=10+9)
CachedTokenAuthenticator/tokens=100_threads=16-60        3.62µs ±17%     0.59µs ±15%  -83.62%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=100_threads=256-60       5.76µs ±25%     0.65µs ±35%  -88.73%  (p=0.000 n=10+9)
CachedTokenAuthenticator/tokens=500_threads=1-60         1.40µs ± 7%     0.38µs ± 2%  -73.14%  (p=0.000 n=9+10)
CachedTokenAuthenticator/tokens=500_threads=16-60        3.96µs ±24%     0.43µs ±17%  -89.15%  (p=0.000 n=9+10)
CachedTokenAuthenticator/tokens=500_threads=256-60       4.47µs ±19%     0.84µs ±21%  -81.16%  (p=0.000 n=10+9)
CachedTokenAuthenticator/tokens=2500_threads=1-60        1.52µs ± 4%     0.36µs ± 1%  -76.10%  (p=0.000 n=8+10)
CachedTokenAuthenticator/tokens=2500_threads=16-60       4.80µs ±19%     0.45µs ±13%  -90.69%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=256-60      3.67µs ±16%     1.22µs ±25%  -66.74%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=1-60       1.94µs ± 5%     0.45µs ± 1%  -77.05%  (p=0.000 n=10+8)
CachedTokenAuthenticator/tokens=12500_threads=16-60      4.06µs ± 8%     0.60µs ± 8%  -85.18%  (p=0.000 n=9+10)
CachedTokenAuthenticator/tokens=12500_threads=256-60     12.0µs ±15%     12.1µs ± 8%     ~     (p=0.436 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=1-60       18.3µs ± 0%     18.0µs ± 0%   -1.62%  (p=0.000 n=10+9)
CachedTokenAuthenticator/tokens=62500_threads=16-60      18.9µs ± 0%     18.8µs ± 0%   -0.99%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=256-60     25.0µs ± 1%     25.0µs ± 1%     ~     (p=0.853 n=10+10)

name                                                  old lookups/op  new lookups/op  delta
CachedTokenAuthenticator/tokens=100_threads=1-60           0.00 ±10%       0.00 ±12%  -44.65%  (p=0.000 n=8+10)
CachedTokenAuthenticator/tokens=100_threads=16-60          0.00 ±44%       0.00 ±49%  -71.53%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=100_threads=256-60         0.03 ±37%       0.01 ±46%  -69.14%  (p=0.000 n=10+9)
CachedTokenAuthenticator/tokens=500_threads=1-60           0.00 ± 9%       0.00 ± 3%  -60.42%  (p=0.000 n=10+9)
CachedTokenAuthenticator/tokens=500_threads=16-60          0.01 ±17%       0.00 ±59%  -86.08%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=500_threads=256-60         0.04 ±23%       0.02 ±20%  -45.23%  (p=0.000 n=10+8)
CachedTokenAuthenticator/tokens=2500_threads=1-60          0.00 ± 8%       0.00 ± 2%  -70.90%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=16-60         0.02 ±41%       0.00 ±24%  -85.10%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=256-60        0.07 ±10%       0.03 ±25%  -47.34%  (p=0.000 n=9+10)
CachedTokenAuthenticator/tokens=12500_threads=1-60         0.03 ± 5%       0.01 ± 1%  -75.01%  (p=0.000 n=10+8)
CachedTokenAuthenticator/tokens=12500_threads=16-60        0.06 ±22%       0.01 ± 7%  -79.81%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=256-60       0.40 ±14%       0.40 ± 7%     ~     (p=0.696 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=1-60         0.66 ± 0%       0.65 ± 0%   -1.69%  (p=0.000 n=10+7)
CachedTokenAuthenticator/tokens=62500_threads=16-60        0.69 ± 0%       0.68 ± 0%   -1.02%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=256-60       0.88 ± 1%       0.88 ± 1%     ~     (p=0.564 n=10+10)

name                                                  old alloc/op    new alloc/op    delta
CachedTokenAuthenticator/tokens=100_threads=1-60           331B ± 0%       300B ± 0%   -9.52%  (p=0.000 n=9+10)
CachedTokenAuthenticator/tokens=100_threads=16-60          346B ± 2%       302B ± 0%  -12.79%  (p=0.000 n=10+7)
CachedTokenAuthenticator/tokens=100_threads=256-60         628B ±17%       345B ± 5%  -45.06%  (p=0.000 n=9+9)
CachedTokenAuthenticator/tokens=500_threads=1-60           329B ± 0%       297B ± 0%   -9.73%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=500_threads=16-60          346B ± 1%       299B ± 0%  -13.64%  (p=0.000 n=9+9)
CachedTokenAuthenticator/tokens=500_threads=256-60         552B ±11%       354B ± 3%  -35.87%  (p=0.000 n=10+9)
CachedTokenAuthenticator/tokens=2500_threads=1-60          331B ± 0%       297B ± 0%  -10.19%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=16-60         352B ± 3%       298B ± 0%  -15.27%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=256-60        582B ± 6%       368B ±10%  -36.68%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=1-60         336B ± 0%       298B ± 0%  -11.31%  (p=0.000 n=9+10)
CachedTokenAuthenticator/tokens=12500_threads=16-60        329B ± 2%       302B ± 0%   -8.29%  (p=0.000 n=10+9)
CachedTokenAuthenticator/tokens=12500_threads=256-60     1.40kB ±13%     1.39kB ± 8%     ~     (p=0.403 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=1-60         537B ± 0%       566B ± 0%   +5.34%  (p=0.000 n=10+8)
CachedTokenAuthenticator/tokens=62500_threads=16-60        630B ± 0%       663B ± 0%   +5.30%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=256-60     2.60kB ± 8%     2.66kB ± 9%     ~     (p=0.165 n=10+10)

name                                                  old allocs/op   new allocs/op   delta
CachedTokenAuthenticator/tokens=100_threads=1-60           13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=100_threads=16-60          13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=100_threads=256-60         14.0 ± 0%       11.0 ± 0%  -21.43%  (p=0.000 n=9+10)
CachedTokenAuthenticator/tokens=500_threads=1-60           13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=500_threads=16-60          13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=500_threads=256-60         13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=1-60          13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=16-60         13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=2500_threads=256-60        14.0 ± 0%       10.4 ± 6%  -25.71%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=1-60         13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=16-60        13.0 ± 0%       11.0 ± 0%  -15.38%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=12500_threads=256-60       16.0 ± 0%       14.0 ± 0%  -12.50%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=1-60         18.0 ± 0%       15.0 ± 0%  -16.67%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=16-60        18.0 ± 0%       15.0 ± 0%  -16.67%  (p=0.000 n=10+10)
CachedTokenAuthenticator/tokens=62500_threads=256-60       20.5 ± 2%       16.0 ± 0%  -21.95%  (p=0.000 n=10+10)
```
#### -benchtime=10s
Long enough to test caching and GC.
```
name                                                  old time/op     new time/op     delta
CachedTokenAuthenticator/tokens=100_threads=1-60         1.54µs ±10%     0.65µs ± 5%   -57.86%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=16-60        3.78µs ± 7%     0.51µs ± 8%   -86.44%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=256-60       5.83µs ± 9%     0.74µs ±15%   -87.25%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=1-60         1.44µs ± 9%     0.37µs ± 1%   -74.10%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=16-60        4.19µs ±17%     0.43µs ± 5%   -89.77%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=256-60       4.73µs ± 9%     0.81µs ±11%   -82.78%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=1-60        1.40µs ± 9%     0.35µs ± 1%   -75.25%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=16-60       4.13µs ±11%     0.42µs ± 7%   -89.88%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=256-60      3.88µs ±11%     0.98µs ±10%   -74.75%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=1-60       1.53µs ± 6%     0.31µs ± 1%   -79.82%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=16-60      3.37µs ±10%     0.49µs ± 9%   -85.35%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=256-60     3.40µs ±11%     1.60µs ± 2%   -52.87%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=1-60       13.9µs ± 0%      0.6µs ± 0%   -95.63%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=16-60      14.0µs ± 0%      0.7µs ± 0%   -94.94%  (p=0.016 n=5+4)
CachedTokenAuthenticator/tokens=62500_threads=256-60     15.7µs ± 0%      4.5µs ± 2%   -71.18%  (p=0.008 n=5+5)

name                                                  old lookups/op  new lookups/op  delta
CachedTokenAuthenticator/tokens=100_threads=1-60           0.00 ± 9%       0.00 ±10%   -32.53%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=16-60          0.00 ±77%       0.00 ±29%   -37.81%  (p=0.032 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=256-60         0.01 ±37%       0.02 ±24%  +311.11%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=1-60           0.00 ±11%       0.00 ± 2%   -54.25%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=16-60          0.00 ±13%       0.00 ± 9%   -61.89%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=256-60         0.01 ±24%       0.02 ±17%  +270.28%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=1-60          0.00 ±13%       0.00 ± 1%   -65.53%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=16-60         0.01 ±23%       0.00 ±21%   -77.04%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=256-60        0.01 ±14%       0.03 ±12%  +147.40%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=1-60         0.01 ±14%       0.00 ± 1%   -77.23%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=16-60        0.02 ±13%       0.01 ±16%   -76.44%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=256-60       0.03 ± 5%       0.05 ± 2%   +61.56%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=1-60         0.50 ± 0%       0.02 ± 0%   -96.96%  (p=0.029 n=4+4)
CachedTokenAuthenticator/tokens=62500_threads=16-60        0.51 ± 0%       0.02 ± 1%   -95.46%  (p=0.016 n=4+5)
CachedTokenAuthenticator/tokens=62500_threads=256-60       0.57 ± 0%       0.16 ± 2%   -71.67%  (p=0.008 n=5+5)

name                                                  old alloc/op    new alloc/op    delta
CachedTokenAuthenticator/tokens=100_threads=1-60           330B ± 0%       299B ± 0%    -9.39%  (p=0.000 n=5+4)
CachedTokenAuthenticator/tokens=100_threads=16-60          326B ± 0%       293B ± 0%   -10.12%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=256-60         365B ± 1%       303B ± 0%   -16.83%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=1-60           329B ± 0%       296B ± 0%   -10.03%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=16-60          330B ± 0%       296B ± 0%      ~     (p=0.079 n=4+5)
CachedTokenAuthenticator/tokens=500_threads=256-60         348B ± 1%       305B ± 0%   -12.57%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=1-60          329B ± 0%       297B ± 0%    -9.73%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=16-60         332B ± 0%       297B ± 0%      ~     (p=0.079 n=4+5)
CachedTokenAuthenticator/tokens=2500_threads=256-60        319B ± 1%       307B ± 0%    -3.70%  (p=0.000 n=5+4)
CachedTokenAuthenticator/tokens=12500_threads=1-60         330B ± 0%       264B ± 0%   -20.00%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=16-60        334B ± 0%       297B ± 0%   -11.18%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=256-60       357B ± 1%       315B ± 0%   -11.93%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=1-60         408B ± 0%       299B ± 0%   -26.72%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=16-60        413B ± 0%       300B ± 0%   -27.36%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=256-60       552B ± 0%       349B ± 0%      ~     (p=0.079 n=4+5)

name                                                  old allocs/op   new allocs/op   delta
CachedTokenAuthenticator/tokens=100_threads=1-60           13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=16-60          13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=100_threads=256-60         13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=1-60           13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=16-60          13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=500_threads=256-60         13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=1-60          13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=16-60         13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=2500_threads=256-60        12.0 ± 0%       11.0 ± 0%    -8.33%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=1-60         13.0 ± 0%       10.0 ± 0%   -23.08%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=16-60        13.0 ± 0%       11.0 ± 0%   -15.38%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=12500_threads=256-60       13.0 ± 0%       12.0 ± 0%    -7.69%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=1-60         16.0 ± 0%       11.0 ± 0%   -31.25%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=16-60        16.0 ± 0%       11.0 ± 0%   -31.25%  (p=0.008 n=5+5)
CachedTokenAuthenticator/tokens=62500_threads=256-60       17.0 ± 0%       12.0 ± 0%   -29.41%  (p=0.008 n=5+5)
```

lavalamp commented:
> `c.heap[0].expiry.After(now)` lines up with the comment better IMO.

Note that `a.After(b)` and `!b.After(a)` are not the same check when a == b. Please do exactly the same check in the Get and gc method; I think the code is correct already (?).

enj commented:
> > `c.heap[0].expiry.After(now)` lines up with the comment better IMO.
> 
> Note that `a.After(b)` and `!b.After(a)` are not the same check when a == b. Please do exactly the same check in the Get and gc method; I think the code is correct already (?).

Yeah that is a fair point.  I just like to avoid negations if possible.  I agree on correctness of existing code.

mikedanese commented:
> Note that `a.After(b)` and `!b.After(a)` are not the same check when a == b. Please do exactly the same check in the Get and gc method; I think the code is correct already (?).

Fixed by only using Before(). Cache entries should not be returned by Get() and should be GC'd if now is equal to or after the expiry of the entry.

mikedanese committed:
inline GC in expiring cache

This allows us to drop the background goroutine with negligable
difference in performance.

mikedanese committed:
fix nasty bug

This needs to be the opposite check of what we do in Get(). Introduced
in a fixup commit.

enj commented:
/lgtm
/retest

liggitt commented:
/approve

mikedanese commented:
Looking at profiles of the stress test, gc only accounts for ~5% of cycles before and after inlining. It's small enough and appears to be fairly distributed between sets so it doesn't make a difference from a performance standpoint. Therefore, not having a background goroutine wins.

mikedanese commented:
/retest

liggitt commented:
/retest

mikedanese commented:
/retest
